### ISSUE TITLE
Add type property to entity documents

### ISSUE THREAD

**Issue by chetanmeh** [apache/incubator-openwhisk#3161]
Currently all whisk entities are stored in "whisks" collection in couchdb. Most queries around these entities are of form [root namespace, date] invoked against entity specific views.

    GET /local_whisks/_design/whisks.v2/_view/packages?descending=true&endkey=["guest",0]&reduce=false&startkey=["guest","￰","￰"]&limit=30
    
The views are computed by "inferring" the entity type [0] based on document properties.This mode of querying works fine with couchdb which provides "computed" views. However it poses problem with storage system where query relies on declarative indexes.

As discussed on [mailing list][1] we should add a `type` property which is set to entity type.

Would open a separate issue for namespace

[0]
```
var isPackage = function (doc) {
        return (doc.binding !== undefined)
 };
var isAction = function (doc) {
     return (doc.exec !== undefined)
 };
 var isTrigger = function (doc) {
        return (doc.exec === undefined && doc.binding === undefined &&
doc.parameters !== undefined)
 };
 var isRule = function (doc) {
        return (doc.trigger !== undefined)
 };
```

[1]: https://lists.apache.org/thread.html/36891e1f3a37312e3923b90717599a4a8a5c8260aea66a0b0fc5f4db@%3Cdev.openwhisk.apache.org%3E

**Comment by chetanmeh** [comment]
@rabbah  Opened PR #3162 with initial approach which uses reflection to determine the type property value based on entity class name. This approach has potential issues like

# Add slight overhead due to use of reflection
# The type value may change in entity class name is refactored later causing backwarnd compatibility issue

Other possible approaches can be

# Add type property explicitly with each type
# OR derive the property in `WhiskDocument` based on same logic as used in CouchDb views

Let me know which approach should be taken

**Comment by rabbah** [comment]
At first glance I think adding type property explicitly is more direct and clean. An abstract method in WhiskEntity that all the subtypes must implement which returns a string and added to the properties written in the WhiskEntity base serializer.

**Comment by chetanmeh** [comment]
Closed via #3162
